providers/frame/polkadot_dev.ts

```diff
@@ -16,15 +30,100 @@ export class PolkadotDevProvider extends FrameBinProvider {
     })
   }
 
+  override async chainFile(pathInfo: PathInfo): Promise<File> {
+    const url = new URL(fromPathInfo({ ...pathInfo, filePath: "test-users" }), this.env.href)
+      .toString()
+    const file = await super.chainFile(pathInfo)
+    return new File(`
+      ${file.codeRaw}
+
+      type ArrayOfLength<
+        T,
+        L extends number,
+        A extends T[] = [],
+      > = number extends L ? T[]
+        : L extends A["length"] ? A
+        : ArrayOfLength<T, L, [...A, T]>
+
+      export async function users<N extends number>(count: N): Promise<ArrayOfLength<C.Sr25519, N>>
+      export async function users(count: number): Promise<C.Sr25519[]> {
+        const response = await fetch(
+          "${url}",
+          {
+            method: "POST",
+            headers: {
+              "content-type": "application/json",
+            },
+            body: JSON.stringify({ count }),
+          },
+        )
+        if (response.status >= 400) {
+          throw new Error(await response.text())
+        }
+        const { from, to }: { from: number; to: number } = await response.json()
+        const userIds: C.Sr25519[] = []
+        for (let i = from; i < to; i++) {
+          userIds.push(C.testUser(i))
+        }
+        return userIds
+      }
+    `)
+  }
+
+  override async handle(request: Request, pathInfo: PathInfo): Promise<Response> {
+    if (request.method.toUpperCase() === "POST" && pathInfo.filePath === "test-users") {
+      try {
+        const { count } = await request.json()
+        if (!(count > 0)) {
+          throw new Error("'count' should be greater than 0")
+        }
+        const currentCount = this.#testUserCountCache[this.#getRuntime(pathInfo)] ?? 0
+        if (count + currentCount > TEST_USER_COUNT) {
+          throw new Error("Maximum test user count reached")
```

Comment:
No need to throw an error on the server. Best to return something that indicates it's out of bounds. The front-end util can then throw.

Reply:
I disagree; throwing a server error for this rare edge case is IMO a lot better than making the return something discriminated and introducing additional complexity.

Reply:
But then we don't see the error from the perspective of the client––but rather from that of the server. I don't think this is preferable; from the server's perspective, there's no error. The server dulls out current test user indices until it reaches the max, at which point it could return a `-1`. The test scripts / clients are responsible for not passing the limit. Ie., they're the ones who step on the land mine / should incur the wrath of a throw.

Reply:
Will we want a custom error on the client side? If so, doesn't make a huge difference whether it's signaled via a 500 or a -1 body. Either way, triggering this error will happen rarely. Fine either way.

Reply:
> Will we want a custom error on the client side?

Not worth the additional complexity IMO. With 10000 if you hit the limit you're doing something very wrong.